src/main/scala/vpio/IO.scala

```diff
@@ -19,6 +21,70 @@ import com.monovore.decline._
 
 // --- //
 
+object Util {
+  import spray.json._
+  import spray.json.DefaultJsonProtocol._
+  import vectorpipe.osm._
+  import vectorpipe.util._
+  import java.io.ByteArrayInputStream
+  import geotrellis.spark.io.s3.S3Client
+  import com.amazonaws.services.s3.model.{ObjectMetadata, AmazonS3Exception}
+
+  implicit object ElementMetaWriter  extends JsonWriter[ElementMeta] {
+    def write(m: ElementMeta): JsValue =
+      JsObject(
+        "id" -> JsNumber(m.id),
+        "user" -> JsString(m.user),
+        "userId" -> JsString(m.userId),
+        "changeSet" -> JsNumber(m.changeSet),
+        "version" -> JsNumber(m.version),
+        "timestamp" -> JsString(m.timestamp),
+        "visible" -> JsBoolean(m.visible)
+      )
+  }
+
+  implicit object ElementDataWriter  extends JsonWriter[ElementData] {
+    def write(d: ElementData): JsValue =
+      JsObject(
+        "meta" -> d.meta.toJson,
+        "tags" -> d.tagMap.toJson
+      )
+  }
+
+  implicit def treeJsonWriter[T:JsonWriter]: JsonWriter[Tree[T]] =
+    new JsonWriter[Tree[T]] {
+      def write(t: Tree[T]): JsValue =
+        if(t.children.isEmpty) { t.root.toJson }
+        else {
+          JsObject(
+            "root" -> t.root.toJson,
+            "children" -> JsArray(t.children.map(write):_*)
+          )
+        }
+    }
+
+  val s3Client: S3Client = S3Client.DEFAULT
+
+  def logClipFail(e: Extent, f: osm.OSMFeature): Unit = {
+    val txt = f.toGeoJson
+    val fid = f.data.root.meta.id
+    val gType =
+      f.geom match {
+        case gc: GeometryCollection => "gc"
+        case mp: MultiPolygon => "mp"
+        case p: Polygon => "p"
+        case ml: MultiLine => "ml"
+        case l: Line => "l"
+        case mp: MultiPoint => "mp"
+        case p: Point => "p"
+      }
+
+    val name = s"finland-run1/${gType}-${fid}.json"
+    val is = new ByteArrayInputStream(txt.getBytes("UTF-8"))
+    s3Client.putObject("vectortiles", s"rde/clip-log/${name}", is, new ObjectMetadata())
```

Comment:
This is a useful way to log, we should codify this in a real thing. 

Also I want to be logging over every step where we skip a thing, e.g. dropping unclosed lines.

Reply:
Agreed for the JSON. Can we use `circe` instead? Spray Json seems to be on its way out.

> Also I want to be logging over every step where we skip a thing, e.g. dropping unclosed lines.

Already implemented :+1:

Reply:
circe is fine by me, just went with what I knew how to do quick

Reply:
Oh but actually - if we have Feature data, we have all our GeoJSON writing stuff in spray, so would have to use spray in this case.